src/string_set.hpp

```diff
@@ -0,0 +1,45 @@
+#include <vector>
+#include <string>
+#include <utility>
+
+#include <cstdint>
+
+struct string_view;
+struct InternedString;
+
+// a flat linear hashing table
+struct StringSet {
+	struct HashField {
+		// possible status values:
+		static constexpr uint64_t Empty = 0;
+		static constexpr uint64_t Occupied = 1;
+		static constexpr uint64_t Tombstone = 2;
+
+		uint64_t status : 2;
+		uint64_t hash_bits : 62;
+	};
+
+	std::vector<std::pair<HashField, std::string>> m_data;
+	size_t m_size;
+
+	StringSet();
+
+	void insert(std::string const&);
+	void insert(std::string&&);
+	// void insert(InternedString const&); // TODO
+	// void insert(string_view const&); // TODO
+	void insert(char const*, size_t);
```

Comment:
Also identical.

Reply:
They are not. the `string&&` overload actually moves the string into the hash table

Reply:
The `string const&` can be implemented as a one liner in terms of the `char*, int` overload, though.